### ISSUE TITLE
no-unused-vars should not consider array write-only methods as 'using' a variable unless the result is used

### ISSUE THREAD

**Issue by gijsk** [eslint/eslint#8414]
**Tell us about your environment**

* **ESLint Version:**
eslint 3.18.0
* **Node Version:**
7.3.0
* **npm Version:**
3.10.10

**What parser (default, Babel-ESLint, etc.) are you using?**
Not sure, presumably the default?

**Please show your full configuration:**

My `.eslintrc.js` for a minimal testcase looks like this:
```js
module.exports = { // eslint-disable-line no-undef
  "rules": {
    "no-unused-vars": ["error"],
  },
};
```

**What did you do? Please include the actual source code causing the issue.**

```js
var foo = [];
foo.push(42);
```

**What did you expect to happen?**
path/to/foo.js
  1:5  error  'foo' is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)

(this is what happens if I comment out the line with the `.push()`)

**What actually happened? Please include the actual, raw output from ESLint.**
No output / errors.

`.push`, like `.splice`, `.shift`, `.unshift`, `.pop`, `.sort` and probably other array methods that I'm forgetting right now, modify the array, but they would only really 'use' the array if their result were used. If nothing else uses the array, `no-unused-vars` should still warn/error as configured.

**Comment by ilyavolodin** [comment]
Unfortunately, we don't know that `foo` is an array. ESLint is a static analysis tool, so we don't know value of the variables at run-time. As such, we can't determine if the function is going to mutate an object, or use it's value. So to be on the safe side, we mark any function usage as usage of the variable.

**Comment by gijsk** [comment]
Well, what's necessary for this to work is the type rather than the specific value, and in some cases that can be determine statically (such as in my example). Does eslint do no type tagging or inference at all? Are there plans to change that?

**Comment by ilyavolodin** [comment]
Type tracking is one of the features we would like to add to ESLint in the future, but no work has been done on this yet. Currently, ESLint can't track types.

**Comment by not-an-aardvark** [comment]
Closing because this is working as intended.
